### PR TITLE
basic implement tso gPRC service

### PR DESCRIPTION
**Pull request by binshi-bing**
Signed-off-by: Bin Shi <binshi.bing@gmail.com>

### What problem does this PR solve?

Issue Number: Ref #5836 

### What is changed and how does it work?

Implement TSO gRPC service. We refactor and move tso rpcs' implementation from pd/server to pkg/tso so that both pd server and the independent TSO mcs can share the same tso logic layer and hide the internal data structures and logics from the upper layer. By doing so, we can adding multi-tenant and sharding while keep grpc layer stable. 

```commit-message
1. Basic implement TSO gRPC service
2. Update kvproto to reuse pdpb tso messages and add GetMembers() rpc. Adjust to the following two main changes in the gPRC protocol:
    a. Add keyspace id and keyspace_group_id to RequestHeader/Response Header to support multi-tenant.
    b. Define new XXXRequest/XXXResponse messages by reusing the messages in pdpb for better back-compatibility. The tenant tso protocol can also evolve independently by adding new rpcs and new messages.
3.  Implement rpcs GetMembers() and Tso()
4. Implement rpcs SyncMaxTS() and GetDCLocationInfo()
5. Implement rpcs SetExternalTimestamp() and GetExternalTimestamp() 

```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/mcs/tso/server/grpc_service.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+// RegisterRESTHandler registers the service to REST server.
+func (s *Service) RegisterRESTHandler(userDefineHandlers map[string]http.Handler) {
+	handler, group := SetUpRestHandler(s)
+	apiutil.RegisterUserDefinedHandlers(userDefineHandlers, &group, handler)
+}
+
+// GetMembers implements gRPC Server.
+func (s *Service) GetMembers(context.Context, *pdpb.GetMembersRequest) (*pdpb.GetMembersResponse, error) {
```
Comment (rleungx) [inline pkg/mcs/tso/server/grpc_service.go:86]: Should we define a new proto for it?

File: server/server.go [hunk 2]
```diff
@@ -1669,37 +1673,12 @@ func (s *Server) RecoverAllocID(ctx context.Context, id uint64) error {
...
-		desc := "the external timestamp should not be larger than global ts"
-		log.Error(desc, zap.Uint64("request timestamp", externalTS), zap.Uint64("global ts", globalTS))
-		return errors.New(desc)
-	}
-	currentExternalTS := s.GetRaftCluster().GetExternalTS()
-	if tsoutil.CompareTimestampUint64(externalTS, currentExternalTS) != 1 {
-		desc := "the external timestamp should be larger than now"
-		log.Error(desc, zap.Uint64("request timestamp", externalTS), zap.Uint64("current external timestamp", currentExternalTS))
-		return errors.New(desc)
-	}
```
Comment (rleungx) [inline server/server.go:1702]: Why remove it?

File: pkg/mcs/tso/server/server.go [hunk 3]
```diff
@@ -60,60 +91,163 @@ func (s *Server) Close() {
...
+
+// GetMember returns the member.
+func (s *Server) GetMember() *member.Member {
+	return s.member
+}
+
+// AddLeaderCallback adds the callback function when the server becomes leader.
+func (s *Server) AddLeaderCallback(callbacks ...func(context.Context)) {
+	s.leaderCallbacks = append(s.leaderCallbacks, callbacks...)
+}
```
Comment (rleungx) [inline pkg/mcs/tso/server/server.go:115]: seem we don't need it?

File: server/server.go (hunk 2 above)
Comment (binshi-bing) [reply server/server.go:1702]: It's moved to pkg/tso

File: pkg/mcs/tso/server/server.go (hunk 3 above)
Comment (binshi-bing) [reply pkg/mcs/tso/server/server.go:160]: @lhy1024 is refining this part.

File: pkg/mcs/tso/server/grpc_service.go (hunk 1 above)
Comment (nolouch) [reply pkg/mcs/tso/server/grpc_service.go:86]: Will using `tsopb` be compatible with previous requests?

File: pkg/mcs/tso/server/server.go (hunk 3 above)
Comment (lhy1024) [reply pkg/mcs/tso/server/server.go:160]: > seem we don't need it?

resource manager server need it

File: pkg/mcs/tso/server/grpc_service.go (hunk 1 above)
Comment (binshi-bing) [reply pkg/mcs/tso/server/grpc_service.go:86]: nope. check my notes in https://github.com/pingcap/kvproto/pull/1053/commits/b64a490910d4da816b39e19803cad2c4f184003e. Also pasted below:

    // GetMembers get the member list of this cluster. It does not require the
    // cluster_id inrequest matchs the id of this cluster. If the server side
    // supports multi-tenant, this rpc only returns the cluster membership info.
    rpc GetMembers(pdpb.GetMembersRequest) returns (pdpb.GetMembersResponse) {}

**Comment by rleungx** [comment]
Shall we close it now?

**Comment by binshi-bing** [comment]
Broken down into separate prs.

**Comment by binshi-bing** [comment]
> Shall we close it now?

closed

**Pull request closed without merging**